src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using System.Web.Hosting;
+using System.Web.Http;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Timers;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public class WebScriptHostExceptionHandler : WebJobExceptionHandler
+    {
+        /// <summary>
+        /// If the exception is a <see cref="TimeoutException"/>, log details and gracefully shut down.
+        /// Otherwise, rethrow the exception from another thread to kill the process.
+        /// </summary>
+        /// <param name="exceptionInfo">The <see cref="ExceptionDispatchInfo"/> to handle.</param>
+        public override void Handle(ExceptionDispatchInfo exceptionInfo)
+        {
+            if (IsTimeout(exceptionInfo))
+            {
+                WebScriptHostManager webScriptHostManager = GlobalConfiguration.Configuration.DependencyResolver.GetService<WebScriptHostManager>();
+                webScriptHostManager?.Instance.TraceWriter.Error("A function timeout has occurred. Host is shutting down.");
+
+                // Give the manager some time to shut down gracefully.
+                Task shutdownTask = Task.Delay(2000).ContinueWith((ct) =>
+                {
+                    HostingEnvironment.InitiateShutdown();
+                });
+
+                webScriptHostManager?.Stop();
```

Comment:
Why don't we stop the host up above the grace period delay? Seems like that is the first thing we should do - we don't want to be taking on new work if we know we're shutting down. We can't block on the Stop call though.